src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializerNamespaces.cs

```diff
@@ -86,51 +84,64 @@ public XmlQualifiedName[] ToArray()
         /// </devdoc>
         public int Count
         {
-            get { return Namespaces.Count; }
+            get { return (_namespaces == null) ? 0 : _namespaces.Count; }
         }
 
-        internal ArrayList? NamespaceList
+        internal IEnumerable<XmlQualifiedName> Namespaces => NamespacesInternal.Values;
+
+        private Dictionary<string, XmlQualifiedName> NamespacesInternal
         {
             get
             {
-                if (_namespaces == null || _namespaces.Count == 0)
-                    return null;
-                ArrayList namespaceList = new ArrayList();
-                foreach (string key in Namespaces.Keys)
-                {
-                    namespaceList.Add(new XmlQualifiedName(key, (string?)Namespaces[key]));
-                }
-                return namespaceList;
+                if (_namespaces == null)
+                    _namespaces = new Dictionary<string, XmlQualifiedName>();
+                return _namespaces;
             }
         }
 
-        [AllowNull]
-        internal Dictionary<string, string?> Namespaces
+        internal ArrayList? NamespaceList
         {
             get
             {
-                if (_namespaces == null)
-                    _namespaces = new Dictionary<string, string?>();
-                return _namespaces;
+                if (_namespaces == null || _namespaces.Count == 0)
+                    return null;
+
+                return new ArrayList(_namespaces.Values);
             }
-            set { _namespaces = value; }
         }
 
-        internal string? LookupPrefix(string? ns)
+        internal bool TryLookupPrefix(string? ns, out string? prefix)
         {
-            if (string.IsNullOrEmpty(ns))
-                return null;
-            if (_namespaces == null || _namespaces.Count == 0)
-                return null;
+            prefix = null;
 
-            foreach (string prefix in _namespaces.Keys)
+            if (string.IsNullOrEmpty(ns) || _namespaces == null || _namespaces.Count == 0)
+                return false;
+
+            foreach (string pfx in _namespaces.Keys)
             {
```

Comment:
Iterate over _namespaces and not _namespaces.Keys so you can avoid the dictionary lookup a couple of lines below where you do `_namespaces[pfx].Namespace`